### PR TITLE
Chore  clean up unused files

### LINKED ISSUE #47971: Clean up unused files and exports
As discussed in https://github.com/freeCodeCamp/freeCodeCamp/pull/47836 there are several unused files and exports.  We should remove all of these from the codebase.

The full set of tasks is:

- [ ] Remove unused files
- [ ] Add missing dependencies
- [ ] Remove unused exports
- [ ] Remove unused types
- [ ] Remove duplicate exports

The files with issues are:

--- UNUSED FILES (8)
src/assets/icons/donate-with-PayPal.tsx
src/assets/icons/intro-information.tsx
src/components/Donation/duration-amount-options.tsx
src/components/Header/components/user-state.tsx
src/components/helpers/form/block-save-wrapper.tsx
src/components/search/searchPage/empty-search.tsx
src/components/search/searchPage/no-results.tsx
src/components/search/searchPage/search-page-hits.tsx

--- UNLISTED DEPENDENCIES (4)
react-instantsearch-core  src/components/search/searchBar/search-bar.tsx
react-instantsearch-core  src/components/search/searchBar/search-hits.tsx
react-instantsearch-core  src/components/search/searchBar/search-suggestion.tsx
gatsby-link               src/templates/Introduction/components/cert-challenge.tsx

--- UNUSED EXPORTS (46)
DonationOptionsText              src/components/Donation/donation-text-components.tsx
DonationSupportText              src/components/Donation/donation-text-components.tsx
PaypalButton                     src/components/Donation/paypal-button.tsx
sagas                            src/components/Flash/redux/index.ts
NavLinks                         src/components/Header/components/nav-links.tsx
searchIndexNameSelector          src/components/search/redux/index.js
updateSearchIndexName            src/components/search/redux/index.js
SearchBarOptimized               src/components/search/searchBar/search-bar-optimized.tsx
SearchBar                        src/components/search/searchBar/search-bar.tsx
CertificationSettings            src/components/settings/certification.js
updateCurrentChallengeId         src/redux/actions.js
initialState                     src/redux/index.js
defaultFetchState                src/redux/index.js
CurrentCertsPropType             src/redux/prop-types.ts
UserPropType                     src/redux/prop-types.ts
saveChallengeSaga                src/redux/save-challenge-saga.js
sessionMetaSelector              src/redux/selectors.js
defaultDonationFormState         src/redux/types.ts
defaultFetchState                src/redux/types.ts
CompletionModalInner             src/templates/Challenges/components/completion-modal.tsx
getCompletedPercent              src/templates/Challenges/components/completion-modal.tsx
HelpModal                        src/templates/Challenges/components/help-modal.tsx
ShortcutsModal                   src/templates/Challenges/components/shortcuts-modal.tsx
VideoModal                       src/templates/Challenges/components/video-modal.tsx
ToolPanel                        src/templates/Challenges/projects/tool-panel.tsx
replaceNBSP                      src/templates/Challenges/rechallenge/transformers.js
testJS$JSX                       src/templates/Challenges/rechallenge/transformers.js
checkChallenge                   src/templates/Challenges/redux/actions.js
lockCode                         src/templates/Challenges/redux/actions.js
updateJSEnabled                  src/templates/Challenges/redux/actions.js
updateSuccessMessageSaga         src/templates/Challenges/redux/current-challenge-saga.js
currentChallengeSaga             src/templates/Challenges/redux/current-challenge-saga.js
executeChallengeSaga             src/templates/Challenges/redux/execute-challenge-saga.js
executeCancellablePreviewSaga    src/templates/Challenges/redux/execute-challenge-saga.js
executeCancellableChallengeSaga  src/templates/Challenges/redux/execute-challenge-saga.js
currentTabSelector               src/templates/Challenges/redux/selectors.js
buildBackendChallenge            src/templates/Challenges/utils/build.ts
buildJSChallenge                 src/templates/Challenges/utils/build.ts
buildDOMChallenge                src/templates/Challenges/utils/build.ts
Block                            src/templates/Introduction/components/block.tsx
CertChallenge                    src/templates/Introduction/components/cert-challenge.tsx
handledErrorSymbol               src/utils/handled-error.ts
wrapHandledError                 src/utils/handled-error.ts
pathParsers                      src/utils/path-parsers.ts
pathsMap                         utils/challenge-types.js
injectConditionalTags            utils/tags.tsx

--- UNUSED TYPES (17)
LoginProps                interface  src/components/Header/components/Login.tsx
AuthOrProfileProps        interface  src/components/Header/components/auth-or-profile.tsx
MenuButtonProps           interface  src/components/Header/components/menu-button.tsx
NavLinksProps             interface  src/components/Header/components/nav-links.tsx
UniversalNavProps         interface  src/components/Header/components/universal-nav.tsx
HeaderProps               interface  src/components/Header/index.tsx
Validator                 type       src/components/formHelpers/form-validators.tsx
FormProps                 type       src/components/formHelpers/form.tsx
FileKeyChallenge          type       src/redux/prop-types.ts
SavedChallenges           type       src/redux/prop-types.ts
ClaimedCertifications     type       src/redux/prop-types.ts
CertTest                  type       src/redux/prop-types.ts
AllMarkdownRemark         type       src/redux/prop-types.ts
Required                  type       src/redux/prop-types.ts
DefaultDonationFormState  interface  src/redux/types.ts
DefaultFetchState         interface  src/redux/types.ts
HandledError              type       src/utils/handled-error.ts

--- DUPLICATE EXPORTS (4)
PayPalButtonScriptLoader, default src/components/Donation/paypal-button-script-loader.tsx
UniversalNav, default             src/components/Header/components/universal-nav.tsx
Header, default                   src/components/Header/index.tsx
Map, default                      src/components/Map/index.tsx

### PR DESCRIPTION
**Pull request by jmetev1**
Checklist:
scope: tools/scripts

- [ x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x ] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [ x] My pull request targets the `main` branch of freeCodeCamp.
- [x ] I have tested these changes either locally on my machine, or GitPod.

Closes #47971 
I'm not sure if it really closes the issue since it seems like it might just be an ongoing chore

### CODE REVIEW

**Comment by jmetev1** [comment]
pnpm knip:all output before:
```Unused files (1)
client/src/assets/icons/spacer.tsx
Unused dependencies (12)
@babel/runtime                  client/package.json
@freecodecamp/strip-comments    client/package.json
@redux-saga/core                client/package.json
@testing-library/jest-dom       client/package.json
assert                          client/package.json
buffer                          client/package.json
eslint                          client/package.json
gatsby-cli                      client/package.json
gatsby-plugin-advanced-sitemap  client/package.json
gatsby-remark-prismjs           client/package.json
sass.js                         client/package.json
invariant                       package.json
Unused dev dependencies (31)
@babel/types                          client/package.json
@codesee/babel-plugin-instrument      client/package.json
@codesee/tracker                      client/package.json
@testing-library/jest-dom             client/package.json
@total-typescript/ts-reset            client/package.json
@types/jest                           client/package.json
@types/testing-library__jest-dom      client/package.json
chokidar                              client/package.json
core-js                               client/package.json
jest-environment-jsdom                client/package.json
@babel/eslint-parser                  package.json
@babel/plugin-proposal-function-bind  package.json
@babel/preset-env                     package.json
@babel/preset-react                   package.json
@testing-library/dom                  package.json
@testing-library/jest-dom             package.json
@testing-library/user-event           package.json
babel-jest                            package.json
babel-plugin-transform-imports        package.json
eslint-plugin-prefer-object-spread    package.json
execa                                 package.json
js-yaml                               package.json
lodash                                package.json
markdownlint                          package.json
mock-fs                               package.json
prismjs                               package.json
process                               package.json
yargs                                 package.json
debug                                 tools/scripts/build/package.json
dotenv                                tools/scripts/build/package.json
joi                                   tools/scripts/build/package.json
Unlisted or unresolved dependencies (7)
@typescript-eslint/eslint-plugin  .eslintrc.json
@typescript-eslint/parser         .eslintrc.json
babel-eslint                      .eslintrc.json
fcc-source-challenges             client/gatsby-config.js
gatsby-remark-node-identity       client/gatsby-config.js
create:config                     client/package.json
jest-environment-jsdom            jest.config.js
Unused exports (7)
searchIndexNameSelector  client/src/components/search/redux/index.js
updateSearchIndexName    client/src/components/search/redux/index.js
SearchBarOptimized       client/src/components/search/searchBar/search-bar-optimized.tsx
query                    client/src/components/seo/index.tsx
completionCountSelector  client/src/redux/selectors.js
emailSelector            client/src/redux/selectors.js
exam                     client/utils/challenge-types.js
Unused exported types (1)
InitTestFrameArg  interface  client/src/client/frame-runner.ts
Unused exported class members (2)
displayName               PayPalButtonScriptLoader  client/src/components/Donation/paypal-button-script-loader.tsx
getDerivedStateFromProps  PayPalButtonScriptLoader  client/src/components/Donation/paypal-button-script-loader.tsx
Duplicate exports (3)
UniversalNav, default  client/src/components/Header/components/universal-nav.tsx
Header, default        client/src/components/Header/index.tsx
Map, default           client/src/components/Map/index.tsx```

**Comment by jmetev1** [comment]
`pnpm knip:all` output after:

```Unlisted or unresolved dependencies (3)
fcc-source-challenges        client/gatsby-config.js
gatsby-remark-node-identity  client/gatsby-config.js
create:config                client/package.json
Unused exports (1)
SearchBarOptimized  client/src/components/search/searchBar/search-bar-optimized.tsx```

File: api/README.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ If you use MongoDB Atlas, the set is managed for you.
 
 The simplest way to run a replica set locally is to use the docker-compose file
 in /tools.
+first install [docker desktop](https://www.docker.com/products/docker-desktop/)
```
Comment (naomi-lgbt) [inline api/README.md:15]: This change is out of scope for this PR.

File: docs/how-to-setup-freecodecamp-locally.md [hunk 2]
```diff
@@ -70,10 +70,10 @@ We primarily support development on Linux and Unix-based systems like Ubuntu and
...
 | [pnpm](https://pnpm.io/installation)                                                          | `8.x`   | -                                                                                            |
-| [MongoDB Community Server](https://docs.mongodb.com/manual/administration/install-community/) | `4.2.x` | -                                                                                           |
+| [MongoDB Community Server](https://docs.mongodb.com/manual/administration/install-community/) | `4.2.x` | [Easiest way is using docker and following](/api/README.md)-                                                                                           |
 
 > [!ATTENTION]
-> If you have a different version, please install the recommended version. We can only support installation issues for recommended versions. See [troubleshooting](#troubleshooting) for details.
+> If you have a different version, please install the recommended version. We can only support installation issues for recommended versions. See [troubleshooting section](troubleshooting-development-issues.md) for details.
```
Comment (naomi-lgbt) [inline docs/how-to-setup-freecodecamp-locally.md:76]: These changes are out of scope for this PR.

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

**Comment by ojeytonwilliams** [comment]
Hi @jmetev1 thanks for your hard work putting this together, but this is too much to review at once.

Also, more importantly, it looks like knip is creating more problems than it's solving. I.e. there are now several things missing that we need.

While I do like the idea behind knip, it has to a) only provide true positives and b) not be too much of a pain to configure correctly. I don't feel like either are true at the moment, at least not for our monorepo.

cc @freeCodeCamp/dev-team for a second opinion, but I'm leaning towards removing `knip` and closing https://github.com/freeCodeCamp/freeCodeCamp/issues/47971

@jmetev1 if you're interested in cleaning up the codebase (it definitely needs it) and have time to create a few smaller, more focused PRs, that would be awesome.

**Comment by jmetev1** [comment]
I'm not gonna lie, that's pretty discouraging. I read the issue and was addressing what it said to address, I read all the contributing guidelines (no mention of PR size), while working I bumped into plenty of other prs that were big. 

Would you like to see it broken up by worksapce or by type of fix (package changes vs removing files) or some other way? @ojeytonwilliams

**Comment by ojeytonwilliams** [comment]
I'm afraid there's no simple rule we can apply to PR size. It mostly comes down to intelligibility. Can a reviewer be reasonably expected to understand precisely what has changed.

I get that in this case the answer should be "nothing", but it's often quite subtle.

Removing package is often very tricky, unfortunately.

So, something that you've done that's super helpful and easy to check is removing unused code. If you're happy to create a PR with just those changes, that would be a great help.
